File: IV3.cpp
#include "IV3.h"
#include "Wire.h"

void IV3_clock::i2c_trans(uint8_t* data, uint8_t length)
{
  Wire.begin();
  Wire.beginTransmission(I2C_address);

  for(int i=0;i<length;i++)
		Wire.write(data[i]);

	Wire.endTransmission();
}

void IV3_clock::set_time(void)
{
  uint8_t TX[7] = {SET_TIME, time.year, time.month, time.day, time.hour, time.minute, time.second};

  i2c_trans(TX, 7);
}

void IV3_clock::read_time(void)
{
  uint8_t RX[6];
  uint8_t TX[1]={READ_TIME};
  uint8_t i;

  i2c_trans(TX, 1);
  delay(10);

  Wire.requestFrom(I2C_address, 6);
  while(Wire.available())
  {
    RX[i] = Wire.read();
    i++;
  }

  time.year = RX[0];
  time.month = RX[1];
  time.day  = RX[2];
  time.hour = RX[3];
  time.minute = RX[4];
  time.second = RX[5];
}

void IV3_clock::clock_mode(void)
{
  uint8_t TX[1]={CLOCK_MODE};
  i2c_trans(TX, 1);
}

void IV3_clock::user_display(uint8_t *data)
{
  uint8_t TX[5];
  uint8_t i;

  TX[0] = USER_DISPLAY;
  for(i=0;i<4;i++)
    TX[i+1] = data[i];

  i2c_trans(TX, 5);
}

void IV3_clock::set_hour_mode(bool mode)
{
  uint8_t TX[2];

  TX[0] = HOUR_MODE;
  if(mode == true)
    TX[1] = 1;
  else
    TX[1] = 0;

  i2c_trans(TX, 2);
}

void IV3_clock::set_brightness(uint8_t brightness)
{
  uint8_t TX[2];

  brightness;

  TX[0] = SET_BRIGHTNESS;
  TX[1] = brightness;

  i2c_trans(TX, 2);
}

void IV3_clock::set_night_fade(bool mode)
{
  uint8_t TX[2];

  TX[0] = NIGHT_FADE;
  if(mode == true)
    TX[1] = 1;
  else
    TX[1] = 0;

  i2c_trans(TX, 2);
}

void IV3_clock::reset(void)
{
  uint8_t TX[1] = {SFT_RST};

  i2c_trans(TX, 1);
}

void IV3_clock::get_status(void)
{
  uint8_t TX[1] = {GET_STATUS};

  i2c_trans(TX, 2);
  delay(10);
  
  Wire.requestFrom(I2C_address, 1);
  while(Wire.available())
  {
    IV3_status = Wire.read();
  }
}

bool IV3_clock::get_hour_mode(void)
{
  this->get_status();
  if((IV3_status & 0x80) == 0x80)
    return true;
  else
    return false;
}

bool IV3_clock::get_fade_mode(void)
{
  this->get_status();
  if((IV3_status & 0x40) == 0x40)
    return true;
  else
    return false;
}

uint8_t IV3_clock::get_brightness(void)
{
  this->get_status();
  IV3_status &= 0x1F;
  return IV3_status;
}

File: CircuitPython/IV3.py
import math
import board
import busio
import time
from digitalio import DigitalInOut, Direction, Pull

class IV3:
    def __init__(self):
        print("Init IV3")
        self.i2c = busio.I2C(board.SCL, board.SDA)
        self.I2C_Address = 0x16
        self.SET_TIME = 0
        self.READ_TIME = 1
        self.CLOCK_MODE = 2
        self.USER_DISPLAY = 3
        self.HOUR_MODE = 4
        self.SET_BRIGHTNESS = 5
        self.NIGHT_FADE = 6
        self.GET_STATUS = 7
        self.SFT_RST = 255

        self.IV3_status = 0

        self.time = {
            "year": 2020,
            "month": 6,
            "day": 28,
            "hour": 16,
            "minute": 41,
            "second": 30
        }

    def _I2C_write(self, data):
        self.i2c.unlock()
        while not self.i2c.try_lock():
            pass
        var = bytearray(data)
        self.i2c.writeto(self.I2C_Address, var)
        self.i2c.unlock()

    def _I2C_read(self, data):
        self.i2c.unlock()
        while not self.i2c.try_lock():
            pass
        self.i2c.readfrom_into(self.I2C_Address, data) ## Read data
        self.i2c.unlock()


    # User methods
    def set_time(self):
        cmd = [self.SET_TIME]
        cmd.append(self.time["year"])
        cmd.append(self.time["month"])
        cmd.append(self.time["day"])
        cmd.append(self.time["hour"])
        cmd.append(self.time["minute"])
        cmd.append(self.time["second"])

        # Write the time to the clock
        self._I2C_write(cmd)

    def read_time(self):
        cmd = [self.READ_TIME]

        self._I2C_write(cmd)

        rx = bytearray(6)
        self._I2C_read(rx)
        
        self.time["year"] = rx[0]
        self.time["month"] = rx[1]
        self.time["day"] = rx[2]
        self.time["hour"] = rx[3]
        self.time["minute"] = rx[4]
        self.time["second"] = rx[5]

    def clock_mode(self):
        self._I2C_write(CLOCK_MODE)

    def user_display(self, data):
        cmd = [self.USER_DISPLAY]

        for i in range(4):
            cmd.append(data[i])
        
        self._I2C_write(cmd)

    def set_hour_mode(self, mode):
        cmd = [self.HOUR_MODE]

        if mode == True:
            cmd.append(1)
        else:
            cmd.append(0)

        self._I2C_write(cmd)

    def set_brightness(self, brightness):
        cmd = [self.SET_BRIGHTNESS]

        if brightness > 20:
            brightness = 20
        elif brightness <= 0:
            brightness = 1

        cmd.append(brightness)

        self._I2C_write(cmd)

    def set_night_fade(self, mode):
        cmd = [self.NIGHT_FADE]

        if mode == True:
            cmd.append(1)
        else:
            cmd.append(0)

        self._I2C_write(cmd)

    def reset(self):
        _I2C_write(self.SFT_RST)

    def get_status(self):
        cmd = [self.GET_STATUS]
        rx = bytearray(1)

        self._I2C_write(cmd)
        self._I2C_read(rx)
        self.IV3_status = rx[0]


    def get_hour_mode(self):   
        
        self.get_status()
        if (self.IV3_status & 0x80) == 0x80:
            return True
        else:
            return False

    def get_fade_mode(self):
        self.get_status()

        if (self.IV3_status & 0x40) == 0x40:
            return True
        else:
            return False

    def get_brightness(self):
        self.get_status()
        return (self.IV3_status & 0x1F)


class numberWang:
    def __init__(self):
        self.switcheroo = {
            1: 0x06,
            2: 0x5B,
            3: 0x4F,
            4: 0x66,
            5: 0x6D,
            6: 0x7D,
            7: 0x07,
            8: 0x7F,
            9: 0x6F,
            0: 0x3F
        }

    def Wanger(self, x):
        wang = [self.switcheroo[math.floor(x / 1000)]]
        wang.append(self.switcheroo[math.floor((x/100)%10)])
        wang.append(self.switcheroo[math.floor((x/10)%10)])
        wang.append(self.switcheroo[math.floor(x % 10)])
        return wang

       